044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,326 @@
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions through the admin api we need to construct an admin client.
+
+If using cluster sourced volumes we require the admin client bootstrap to be configured using `client.quota.callback.kafka.admin.bootstrap.servers`
+
+Additional admin client configuration can be passed using the form `client.quota.callback.kafka.admin.${configuration}`.
+For example: `client.quota.callback.kafka.admin.security.protocol=SSL`
+
+The strimzi-kafka-operator would be responsible for configuring the admin client to connect to the
+replication listener of the broker.
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+used bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume.
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine an availableBytesBelow soft limit with a availableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+2. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+For example, to configure a soft limit when availableBytes is below 5GB and hard limit when availableBytes is below 1GB:
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.soft=5000000000`
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+
+Another example, to configure a soft limit when availableRatio is below 0.05 (5%) and hard limit when availableRatio is below 0.01 (1%):
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.01`
+
+Or mixing the two types, which is also valid to soft limit at 0.05 (5%) availableRatio and hard limit at 1GB availableBytes:
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+
+Defining more than one hard or soft limit is **invalid**, for example the below is disallowed:
```

Comment:
Should we throw an error or apply the [smallest|largest] limit?

Reply:
I'd go for an error if we are disallowing these combinations.

But yeah, it's another option to allow multiple hard/soft limits. Soft limit at 3% available or 5G available whatever you hit first.

Reply:
Sam and I decided to start with disallowing, we could make it less harsh in future